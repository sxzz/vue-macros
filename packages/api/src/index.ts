export * from '@vue-macros/common'

export * from './error'
export * from './vue'
export * from './ts'
