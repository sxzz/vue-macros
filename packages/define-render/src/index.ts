import {
  createFilter,
  detectVueVersion,
  REGEX_SETUP_SFC,
  REGEX_VUE_SFC,
  type BaseOptions,
  type MarkRequired,
} from '@vue-macros/common'
import { generatePluginName } from '#macros' with { type: 'macro' }
import { createUnplugin, type UnpluginInstance } from 'unplugin'
import { transformDefineRender } from './core'

export type Options = BaseOptions & {
  vapor?: boolean
}
export type OptionsResolved = MarkRequired<Options, 'include' | 'version'>

function resolveOptions(options: Options): OptionsResolved {
  const version = options.version || detectVueVersion()
  return {
    include: [
      REGEX_VUE_SFC,
      REGEX_SETUP_SFC,
      /\.(vue|setup\.[cm]?[jt]sx?)\?vue/,
    ],
    version,
    ...options,
  }
}
const name = generatePluginName()

const plugin: UnpluginInstance<Options | undefined, false> = createUnplugin(
  (userOptions = {}) => {
    const options = resolveOptions(userOptions)
    const filter = createFilter(options)

    return {
      name,
      enforce: 'post',

      transformInclude(id) {
        return filter(id)
      },

      transform(code, id) {
        return transformDefineRender(code, id, options)
      },
    }
  },
)
export default plugin
