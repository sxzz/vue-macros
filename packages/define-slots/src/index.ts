import {
  createFilter,
  detectVueVersion,
  FilterFileType,
  getFilterPattern,
  type BaseOptions,
  type MarkRequired,
} from '@vue-macros/common'
import { generatePluginName } from '#macros' with { type: 'macro' }
import {
  createUnplugin,
  type UnpluginContextMeta,
  type UnpluginInstance,
} from 'unplugin'
import { transformDefineSlots } from './core'

export type Options = BaseOptions
export type OptionsResolved = MarkRequired<Options, 'include' | 'version'>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework'],
): OptionsResolved {
  const version = options.version || detectVueVersion()
  const include = getFilterPattern(
    [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SETUP_SFC],
    framework,
  )
  return {
    include,
    ...options,
    version,
  }
}

const name = generatePluginName()

const plugin: UnpluginInstance<Options | undefined, false> = createUnplugin(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)

    return {
      name,
      enforce: 'pre',
      transformInclude: filter,
      transform: transformDefineSlots,
    }
  },
)
export default plugin
