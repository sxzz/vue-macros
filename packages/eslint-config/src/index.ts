/* eslint perfectionist/sort-objects: "error" */

import type { Rules } from '@sxzz/eslint-config'
import type { Linter } from 'eslint'

const globals: Linter.Globals = {
  $: 'readonly',
  $$: 'readonly',
  $computed: 'readonly',
  $customRef: 'readonly',
  $defineModels: 'readonly',
  $defineProps: 'readonly',
  $definePropsRefs: 'readonly',
  $ref: 'readonly',
  $shallowRef: 'readonly',
  $toRef: 'readonly',
  defineEmit: 'readonly',
  defineModels: 'readonly',
  defineOptions: 'readonly',
  defineProp: 'readonly',
  defineProps: 'readonly',
  defineRender: 'readonly',
  defineSetupComponent: 'readonly',
  defineSlots: 'readonly',
}

const rules: Linter.RulesRecord = {
  'vue/no-export-in-script-setup': 'off', // exportRender / exportExpose / exportProps
  'vue/valid-attribute-name': 'off', // short-vmodel
  'vue/valid-define-props': 'off', // hoistStatic
  'vue/valid-v-bind': 'off', // shortBind + shortVmodel
} satisfies Rules

const config: Linter.Config = {
  languageOptions: {
    globals,
  },
  name: 'vue-macros/rules',
  rules,
}

export default config
