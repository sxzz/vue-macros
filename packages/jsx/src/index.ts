import jsxDirective from '@vue-macros/jsx-directive'
import jsxMacros from '@vue-macros/jsx-macros'

import { generatePluginName } from '#macros' with { type: 'macro' }
import {
  createCombinePlugin,
  type UnpluginCombineInstance,
} from 'unplugin-combine'
import { resolveJSXOptions, type JSXOptions } from './options'

const name = generatePluginName()
const plugin: UnpluginCombineInstance<JSXOptions | undefined> =
  createCombinePlugin<JSXOptions | undefined>((userOptions = {}, meta) => {
    const options = resolveJSXOptions(userOptions)
    const framework = meta.framework!
    const plugins = [
      options.directive
        ? jsxDirective[framework](options.directive)
        : undefined,
      options.macros ? jsxMacros[framework](options.macros) : undefined,
    ].filter((plugin) => !!plugin)

    return {
      name,
      plugins,
    }
  })

export default plugin
