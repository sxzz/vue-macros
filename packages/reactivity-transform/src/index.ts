import {
  createFilter,
  createRollupFilter,
  detectVueVersion,
  FilterFileType,
  getFilterPattern,
  normalizePath,
  REGEX_NODE_MODULES,
  type BaseOptions,
  type MarkRequired,
} from '@vue-macros/common'
import { generatePluginName } from '#macros' with { type: 'macro' }
import {
  createUnplugin,
  type UnpluginContextMeta,
  type UnpluginInstance,
} from 'unplugin'
import { shouldTransform, transform, transformVueSFC } from './core'
import { helperCode, helperId } from './core/helper'

export type Options = BaseOptions
export type OptionsResolved = MarkRequired<Options, 'include' | 'version'>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework'],
): OptionsResolved {
  const version = options.version || detectVueVersion()
  const include = getFilterPattern(
    [FilterFileType.SRC_FILE, FilterFileType.VUE_SFC_WITH_SETUP],
    framework,
  )
  return {
    include,
    exclude: [REGEX_NODE_MODULES],
    ...options,
    version,
  }
}

const name = generatePluginName()

const plugin: UnpluginInstance<Options | undefined, false> = createUnplugin(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)
    const filterSFC = createRollupFilter(
      getFilterPattern(
        [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SETUP_SFC],
        framework,
      ),
    )
    return {
      name,
      enforce: 'pre',

      resolveId(id) {
        if (id === normalizePath(helperId)) return id
      },

      loadInclude(id) {
        return normalizePath(id) === helperId
      },

      load(id) {
        if (normalizePath(id) === helperId) return helperCode
      },

      transformInclude: filter,
      transform(code, id) {
        if (filterSFC(id)) {
          return transformVueSFC(code, id)
        } else if (shouldTransform(code)) {
          return transform(code, {
            filename: id,
            sourceMap: true,
          })
        }
      },
    }
  },
)
export default plugin
