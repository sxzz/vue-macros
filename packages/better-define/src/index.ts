import process from 'node:process'
import { resolveDtsHMR } from '@vue-macros/api'
import {
  createFilter,
  detectVueVersion,
  FilterFileType,
  getFilterPattern,
  type BaseOptions,
  type MarkRequired,
} from '@vue-macros/common'
import { generatePluginName } from '#macros' with { type: 'macro' }
import {
  createUnplugin,
  type UnpluginContextMeta,
  type UnpluginInstance,
} from 'unplugin'
import { transformBetterDefine } from './core'

export type Options = BaseOptions
export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'isProduction'
>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework'],
): OptionsResolved {
  const version = options.version || detectVueVersion()
  const include = getFilterPattern(
    [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SETUP_SFC],
    framework,
  )
  return {
    include,
    isProduction: process.env.NODE_ENV === 'production',
    ...options,
    version,
  }
}

const name = generatePluginName()

const plugin: UnpluginInstance<Options | undefined, false> = createUnplugin(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)

    return {
      name,
      enforce: 'pre',

      transformInclude: filter,
      async transform(code, id) {
        try {
          return (
            await transformBetterDefine(code, id, options.isProduction)
          ).match(
            (res) => res,
            (error) => console.warn(error),
          )
        } catch (error: unknown) {
          console.warn(error)
        }
      },

      vite: {
        configResolved(config) {
          options.isProduction ??= config.isProduction
        },

        handleHotUpdate: resolveDtsHMR,
      },
    }
  },
)
export default plugin
