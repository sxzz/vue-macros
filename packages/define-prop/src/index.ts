import process from 'node:process'
import { resolveDtsHMR } from '@vue-macros/api'
import {
  createFilter,
  detectVueVersion,
  FilterFileType,
  getFilterPattern,
  normalizePath,
  type BaseOptions,
  type MarkRequired,
} from '@vue-macros/common'
import { generatePluginName } from '#macros' with { type: 'macro' }
import {
  createUnplugin,
  type UnpluginContextMeta,
  type UnpluginInstance,
} from 'unplugin'
import { transformDefineProp, type Edition } from './core'
import { helperCode, helperId } from './core/helper'

export interface Options extends BaseOptions {
  edition?: Edition
}

export type OptionsResolved = MarkRequired<
  Options,
  'include' | 'version' | 'isProduction' | 'edition'
>

function resolveOptions(
  options: Options,
  framework: UnpluginContextMeta['framework'],
): OptionsResolved {
  const version = options.version || detectVueVersion()
  const include = getFilterPattern(
    [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SETUP_SFC],
    framework,
  )
  return {
    include,
    isProduction: process.env.NODE_ENV === 'production',
    edition: 'kevinEdition',
    ...options,
    version,
  }
}

const name = generatePluginName()

const plugin: UnpluginInstance<Options | undefined, false> = createUnplugin(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework)
    const filter = createFilter(options)

    return {
      name,
      enforce: 'pre',

      resolveId(id) {
        if (id === normalizePath(helperId)) return id
      },

      loadInclude(id) {
        return normalizePath(id) === helperId
      },

      load(id) {
        if (normalizePath(id) === helperId) return helperCode
      },

      transformInclude: filter,
      transform(code, id) {
        return transformDefineProp(
          code,
          id,
          options.edition,
          options.isProduction,
        )
      },

      vite: {
        configResolved(config) {
          options.isProduction ??= config.isProduction
        },

        handleHotUpdate: resolveDtsHMR,
      },
    }
  },
)
export default plugin
