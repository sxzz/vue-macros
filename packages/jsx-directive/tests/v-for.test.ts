import { testFixtures } from '@vue-macros/test-utils'
import { describe } from 'vitest'
import { transformJsxDirective } from '../src/api'

describe('jsx-vue-directive', () => {
  describe('vue 2.7 v-for', async () => {
    await testFixtures(
      import.meta.glob<string>('./fixtures/v-for/*.{vue,jsx,tsx}', {
        eager: true,
        query: '?raw',
        import: 'default',
      }),
      (_, id, code) =>
        transformJsxDirective(code, id, {
          version: 2.7,
          lib: 'vue',
          prefix: 'v-',
        })?.code,
    )
  })

  describe('vue 3 v-for', async () => {
    await testFixtures(
      import.meta.glob<string>('./fixtures/v-for/*.{vue,jsx,tsx}', {
        eager: true,
        query: '?raw',
        import: 'default',
      }),
      (_, id, code) =>
        transformJsxDirective(code, id, {
          version: 3,
          lib: 'vue',
          prefix: 'v-',
        })?.code,
    )
  })

  describe('vapor v-for', async () => {
    await testFixtures(
      import.meta.glob<string>('./fixtures/v-for/*.{vue,jsx,tsx}', {
        eager: true,
        query: '?raw',
        import: 'default',
      }),
      (_, id, code) =>
        transformJsxDirective(code, id, {
          version: 3,
          lib: 'vapor',
          prefix: 'v-',
        })?.code,
    )
  })
})
