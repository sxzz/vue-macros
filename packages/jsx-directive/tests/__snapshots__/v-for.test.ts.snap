// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`jsx-vue-directive > vapor v-for > ./fixtures/v-for/index.vue 1`] = `
"<script setup lang="tsx">
import { createFor as __MACROS_createFor } from "vue";
const map = new Map([
  [1, '2'],
  [3, '4'],
])
const set = new Set(['1', '2', '3'])
const object = { id: 1, name: 'admin' }
let selected = 0

const Comp = <>{__MACROS_createFor(() => 4, (__MACROS_props) => <div>{__MACROS_props[0].value}</div>)}</>

defineRender(() => (
  <>
    {__MACROS_createFor(() => 4, (__MACROS_props) => <div>
      <div>{__MACROS_props[0].value}</div>
    </div>, ([i]) => (i))}

    {__MACROS_createFor(() => object, (__MACROS_props) => <div>
      {__MACROS_props[1].value}: {__MACROS_props[0].value}
    </div>, ([value, key, index]) => (index))}

    {__MACROS_createFor(() => [1, 2, 3][Symbol.iterator](), (__MACROS_props) => <div>
      <div>{__MACROS_props[0].value}</div>
    </div>, ([i, index]) => (index))}

    {__MACROS_createFor(() => [1, 2, 3], (__MACROS_props) => <div v-memo={[selected === __MACROS_props[0].value]}>
      <div>{__MACROS_props[0].value}</div>
    </div>, ([i, __MACROS_index]) => (i))}

    {(set) ? __MACROS_createFor(() => set, (__MACROS_props) => <div>
      <div>{__MACROS_props[0].value}</div>
    </div>, ([i]) => (i)) : null}

    {__MACROS_createFor(() => map, (__MACROS_props) => <div>
      {__MACROS_props[0].value[0].value}
      {__MACROS_createFor(() => 3, (__MACROS_props) => <div>
        {__MACROS_props[0].value}: {__MACROS_props[0].value[1].value}
      </div>)}
    </div>, ([[key, value], index]) => (index))}
  </>
))
</script>
"
`;

exports[`jsx-vue-directive > vue 2.7 v-for > ./fixtures/v-for/index.vue 1`] = `
"<script setup lang="tsx">
const map = new Map([
  [1, '2'],
  [3, '4'],
])
const set = new Set(['1', '2', '3'])
const object = { id: 1, name: 'admin' }
let selected = 0

const Comp = <>{Array.from(4, (i) => <div>{i}</div>)}</>

defineRender(() => (
  <>
    {Array.from(4, (i) => <div key={i}>
      <div>{i}</div>
    </div>)}

    {Array.from(object, (value, key, index) => <div key={index}>
      {key}: {value}
    </div>)}

    {Array.from([1, 2, 3][Symbol.iterator](), (i, index) => <div key={index}>
      <div>{i}</div>
    </div>)}

    {Array.from([1, 2, 3], (i, __MACROS_index) => <div v-memo={[selected === i]} key={i}>
      <div>{i}</div>
    </div>)}

    {(set) ? Array.from(set, (i) => <div key={i}>
      <div>{i}</div>
    </div>) : null}

    {Array.from(map, ([key, value], index) => <div key={index}>
      {key}
      {Array.from(3, (key) => <div>
        {key}: {value}
      </div>)}
    </div>)}
  </>
))
</script>
"
`;

exports[`jsx-vue-directive > vue 3 v-for > ./fixtures/v-for/index.vue 1`] = `
"<script setup lang="tsx">
import { renderList as __MACROS_renderList } from "vue";
const map = new Map([
  [1, '2'],
  [3, '4'],
])
const set = new Set(['1', '2', '3'])
const object = { id: 1, name: 'admin' }
let selected = 0

const Comp = <>{__MACROS_renderList(4, (i) => <div>{i}</div>)}</>

defineRender(() => (
  <>
    {__MACROS_renderList(4, (i) => <div key={i}>
      <div>{i}</div>
    </div>)}

    {__MACROS_renderList(object, (value, key, index) => <div key={index}>
      {key}: {value}
    </div>)}

    {__MACROS_renderList([1, 2, 3][Symbol.iterator](), (i, index) => <div key={index}>
      <div>{i}</div>
    </div>)}

    {__MACROS_renderList([1, 2, 3], (i, __MACROS_index) => <div v-memo={[selected === i]} key={i}>
      <div>{i}</div>
    </div>)}

    {(set) ? __MACROS_renderList(set, (i) => <div key={i}>
      <div>{i}</div>
    </div>) : null}

    {__MACROS_renderList(map, ([key, value], index) => <div key={index}>
      {key}
      {__MACROS_renderList(3, (key) => <div>
        {key}: {value}
      </div>)}
    </div>)}
  </>
))
</script>
"
`;

exports[`jsx-vue-directive > vue/vapor v-for > ./fixtures/v-for/index.vue 1`] = `
"<script setup lang="tsx">
import { createFor as __MACROS_createFor } from "vue";
const map = new Map([
  [1, '2'],
  [3, '4'],
])
const set = new Set(['1', '2', '3'])
const object = { id: 1, name: 'admin' }
let selected = 0

const Comp = <>{__MACROS_createFor(() => 4, (__MACROS_props) => <div>{__MACROS_props[0].value}</div>)}</>

defineRender(() => (
  <>
    {__MACROS_createFor(() => 4, (__MACROS_props) => <div>
      <div>{__MACROS_props[0].value}</div>
    </div>, ([i]) => (i))}

    {__MACROS_createFor(() => object, (__MACROS_props) => <div>
      {__MACROS_props[1].value}: {__MACROS_props[0].value}
    </div>, ([value, key, index]) => (index))}

    {__MACROS_createFor(() => [1, 2, 3][Symbol.iterator](), (__MACROS_props) => <div>
      <div>{__MACROS_props[0].value}</div>
    </div>, ([i, index]) => (index))}

    {__MACROS_createFor(() => [1, 2, 3], (__MACROS_props) => <div v-memo={[selected === __MACROS_props[0].value]}>
      <div>{__MACROS_props[0].value}</div>
    </div>, ([i, __MACROS_index]) => (i))}

    {(set) ? __MACROS_createFor(() => set, (__MACROS_props) => <div>
      <div>{__MACROS_props[0].value}</div>
    </div>, ([i]) => (i)) : null}

    {__MACROS_createFor(() => map, (__MACROS_props) => <div>
      {__MACROS_props[0].value[0].value}
      {__MACROS_createFor(() => 3, (__MACROS_props) => <div>
        {__MACROS_props[0].value}: {__MACROS_props[0].value[1].value}
      </div>)}
    </div>, ([[key, value], index]) => (index))}
  </>
))
</script>
"
`;
